curvefs/src/mds/topology/topology.cpp

```diff
@@ -1065,35 +1065,72 @@ int TopologyImpl::GetOneRandomNumber(int start, int end) const {
     return dis(gen);
 }
 
-TopoStatusCode TopologyImpl::ChooseSinglePoolRandom(PoolIdType *out) const {
+TopoStatusCode TopologyImpl::ChooseSinglePoolRandom(PoolIdType *out,
+    const std::set<PoolIdType> &unavailablePools) const {
     ReadLockGuard rlockPool(poolMutex_);
-    if (poolMap_.empty()) {
-        LOG(ERROR) << "Choose single pool failed, poolmap is empty.";
+    std::unordered_map<PoolIdType, Pool> tmpPoolMap = poolMap_;
+
+    // erase unavailable pools
+    for (const auto &pool : unavailablePools) {
+        auto it = tmpPoolMap.find(pool);
+        if (it != tmpPoolMap.end()) {
+            tmpPoolMap.erase(it);
+        }
+    }
+
+    if (tmpPoolMap.empty()) {
+        LOG(ERROR) << "Choose single pool failed, have no available pool.";
         return TopoStatusCode::TOPO_POOL_NOT_FOUND;
     }
-    int randomValue = GetOneRandomNumber(0, poolMap_.size() - 1);
-    auto iter = poolMap_.begin();
+
+    int randomValue = GetOneRandomNumber(0, tmpPoolMap.size() - 1);
+    auto iter = tmpPoolMap.begin();
     std::advance(iter, randomValue);
     *out = iter->second.GetId();
     return TopoStatusCode::TOPO_OK;
 }
 
 TopoStatusCode TopologyImpl::ChooseZonesInPool(PoolIdType poolId,
-                                               std::set<ZoneIdType> *zones,
-                                               int count) const {
+    std::set<ZoneIdType> *zones,
+    const std::set<ZoneIdType> &unavailableZones,
+    int count) const {
     std::list<ZoneIdType> zoneList = GetZoneInPool(poolId);
-    if (zoneList.size() < count) {
+    if (zoneList.size() < count + zones->size()) {
         LOG(ERROR) << "Choose zone in pool failed,"
                    << "the zone.size = " << zoneList.size()
-                   << ", the need count = " << count;
+                   << ", the need count = " << count
+                   << ", the unavailableZone = " << unavailableZones.size();
         return TopoStatusCode::TOPO_ZONE_NOT_FOUND;
     }
 
-    while (zones->size() < count) {
+    // erase unavailable zone
+    for (const auto &zone : unavailableZones) {
+        auto it = std::find(zoneList.begin(), zoneList.end(), zone);
+        if (it != zoneList.end()) {
+            zoneList.erase(it);
+        }
+        auto iter = zones->find(zone);
+        if (iter != zones->end()) {
+            zones->erase(iter);
+        }
+    }
+    // erase already get zones
+    auto it = zones->begin();
+    while (it != zones->end()) {
+        auto iter = std::find(zoneList.begin(), zoneList.end(), *it);
+        if (iter != zoneList.end()) {
+            zoneList.erase(iter);
+        }
+        it++;
+    }
+
+    while (count > 0) {
```

Comment:
you should check whether `zoneList.size() >= count`

Reply:
> you should check whether `zoneList.size() >= count`

It checked in line1098.